library/src/com/google/maps/android/kml/KmlRenderer.java

```diff
@@ -0,0 +1,870 @@
+     * @param geometry defines the type of object to add to the map
+     * @param style    defines styling properties to add to the object when added to the map
+     * @return the object that was added to the map, this is a Marker, Polyline, Polygon or an array
+     * of either objects
+     */
+    private Object addToMap(KmlPlacemark placemark, KmlGeometry geometry, KmlStyle style,
+            KmlStyle inlineStyle, Boolean isVisible) {
+        String geometryType = geometry.getKmlGeometryType();
+        if (geometryType.equals("Point")) {
+            Marker marker = addPointToMap(placemark, (KmlPoint) geometry, style, inlineStyle);
+            marker.setVisible(isVisible);
+            return marker;
+        } else if (geometryType.equals("LineString")) {
+            Polyline polyline = addLineStringToMap((KmlLineString) geometry, style, inlineStyle);
+            polyline.setVisible(isVisible);
+            return polyline;
+        } else if (geometryType.equals("Polygon")) {
+            Polygon polygon = addPolygonToMap((KmlPolygon) geometry, style, inlineStyle);
+            polygon.setVisible(isVisible);
+            return polygon;
+        } else if (geometryType.equals("MultiGeometry")) {
+            return addMultiGeometryToMap(placemark, (KmlMultiGeometry) geometry, style, inlineStyle,
+                    isVisible);
+        }
+        return null;
+    }
+
+    /**
+     * Adds a KML Point to the map as a Marker by combining the styling and coordinates
+     *
+     * @param point contains coordinates for the Marker
+     * @param style contains relevant styling properties for the Marker
+     * @return Marker object
+     */
+    private Marker addPointToMap(KmlPlacemark placemark, KmlPoint point, KmlStyle style,
+            KmlStyle markerInlineStyle) {
+        MarkerOptions markerUrlStyle = style.getMarkerOptions();
+        markerUrlStyle.position(point.getKmlGeometryObject());
+        if (markerInlineStyle != null) {
+            setInlinePointStyle(markerUrlStyle, markerInlineStyle, style.getIconUrl());
+        } else if (style.getIconUrl() != null) {
+            // Use shared style
+            addMarkerIcons(style.getIconUrl(), markerUrlStyle);
+        }
+        Marker marker = mMap.addMarker(markerUrlStyle);
+        setMarkerInfoWindow(style, marker, placemark);
+        return marker;
+    }
+
+    /**
+     * Sets a marker info window if no <text> tag was found in the KML document. This method sets
+     * the marker title as the text found in the <name> start tag and the snippet as <description>
+     *
+     * @param style Style to apply
+     */
+    private void setMarkerInfoWindow(KmlStyle style, Marker marker,
+            KmlPlacemark placemark) {
+        Boolean hasName = placemark.getProperty("name") != null;
+        Boolean hasDescription = placemark.getProperty("description") != null;
+        if (style.getBalloonOptions() != null && style.getBalloonOptions().containsKey("text")) {
```

Comment:
Since we are adding Boolean values like hasName and hasDescription, for consistency we should assign hasBalloonOptions and hasBalloonText in this line